# LAB __ Nivel 0

In [3]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi -c "SELECT 1;"

 ?column? 
----------
        1
(1 row)



# FASE 1 Capa RAW

Lectura del fichero parquet y transformación a fichero CSV (Se necesita importar pandas para poder transformar de parquet a CSV)

In [1]:
import pandas as pd
df = pd.read_parquet('yellow_tripdata_2024-01.parquet')
df.to_csv('yellow_tripdata_2024-01.csv',index=False)

Creación de un nuevo esquema RAW (BRONZE)

In [4]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE SCHEMA IF NOT EXISTS raw;
SQL

CREATE SCHEMA


Creación de la tabla RAW yellow_taxi_raw (Si nos fijamos todas la columnas se quedan en formato texto, no hacemos suposiciones)

In [5]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
DROP TABLE IF EXISTS raw.yellow_taxi_raw;

CREATE TABLE raw.yellow_taxi_raw (
    vendor_id TEXT,
    pickup_datetime TEXT,
    dropoff_datetime TEXT,
    passenger_count TEXT,
    trip_distance TEXT,
    ratecode_id TEXT,
    store_and_fwd_flag TEXT,
    pulocation_id TEXT,
    dolocation_id TEXT,
    payment_type TEXT,
    fare_amount TEXT,
    extra TEXT,
    mta_tax TEXT,
    tip_amount TEXT,
    tolls_amount TEXT,
    improvement_surcharge TEXT,
    total_amount TEXT,
    congestion_surcharge TEXT,
    airport_fee TEXT
);
SQL

NOTICE:  table "yellow_taxi_raw" does not exist, skipping


DROP TABLE
CREATE TABLE


Carga CSV a la tabla RAW "yellow.taxi.raw"

In [8]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
\copy raw.yellow_taxi_raw FROM 'yellow_tripdata_2024-01.csv' CSV HEADER;
SQL

COPY 2964624


Comprobamos la tabla con datos

In [9]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
SELECT COUNT(*) FROM raw.yellow_taxi_raw;
SQL

  count  
---------
 2964624
(1 row)



# FASE 2 - Modelo relacional (CURATED)

Creamos el esquema CURATED (SILVER)

In [10]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE SCHEMA IF NOT EXISTS curated;
SQL

CREATE SCHEMA


Creamos la tabla TRIP 

In [11]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
DROP TABLE IF EXISTS curated.trip CASCADE;

CREATE TABLE curated.trip (
    trip_id BIGSERIAL PRIMARY KEY,
    vendor_id INTEGER,
    pickup_datetime TIMESTAMP,
    dropoff_datetime TIMESTAMP,
    passenger_count INTEGER,
    trip_distance NUMERIC,
    fare_amount NUMERIC,
    tip_amount NUMERIC,
    total_amount NUMERIC,
    pulocation_id INTEGER,
    dolocation_id INTEGER
);
SQL

NOTICE:  table "trip" does not exist, skipping


DROP TABLE
CREATE TABLE


Pasamos la información de la capa RAW a la capa CURATED

In [1]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
INSERT INTO curated.trip (
    vendor_id,
    pickup_datetime,
    dropoff_datetime,
    passenger_count,
    trip_distance,
    fare_amount,
    tip_amount,
    total_amount,
    pulocation_id,
    dolocation_id
)
SELECT
    vendor_id::INTEGER,
    pickup_datetime::TIMESTAMP,
    dropoff_datetime::TIMESTAMP,
    passenger_count::NUMERIC::INTEGER,
    trip_distance::NUMERIC,
    fare_amount::NUMERIC,
    tip_amount::NUMERIC,
    total_amount::NUMERIC,
    pulocation_id::INTEGER,
    dolocation_id::INTEGER
FROM raw.yellow_taxi_raw;
SQL

INSERT 0 2964624


Comprobamos el volumen de la tabla curated

In [2]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
SELECT COUNT(*) FROM curated.trip;
SQL

  count  
---------
 2964624
(1 row)



# FASE 3 ANALYTICS -- EDA Indices y Performance (Gold)

In [4]:
%load_ext sql
%sql postgresql://root:root@localhost:5432/ny_taxi

Connecting to 'postgresql://root:***@localhost:5432/ny_taxi'

EDA basico

In [4]:
%%sql
SELECT
  COUNT(*) AS total_trips,
  MIN(pickup_datetime),
  MAX(pickup_datetime)
FROM
  curated.trip;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

1 rows affected.

total_trips,min,max
2964624,2002-12-31 22:59:39,2024-02-01 00:01:15


problema de calidad de datos heredado de la fuente: el csv corresponde a 2024 pero el pickup-datetime es 2002-12-31.

Ingresos por día

In [5]:
%%sql
SELECT
  DATE (pickup_datetime) AS DAY,
  SUM(total_amount) AS revenue
FROM
  curated.trip
GROUP BY
  DAY
ORDER BY
  DAY;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

35 rows affected.

day,revenue
2002-12-31,0.0
2009-01-01,127.69
2023-12-31,224.62
2024-01-01,2442843.25
2024-01-02,2282196.02
2024-01-03,2357588.48
2024-01-04,2800511.06
2024-01-05,2728672.52
2024-01-06,2436272.32
2024-01-07,1898102.16


Indice para acelerar la busqueda y filtrado de filas. creamos un indice de pickup_datetime para optimizar consultas analiticas basadas en rangos de tiempo.

In [6]:
%%sql
CREATE INDEX idx_trip_pickup_datetime ON curated.trip (pickup_datetime);

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

++
||
++
++

Ahora creamos un explain: 

In [7]:
%%sql
EXPLAIN ANALYZE
SELECT
  *
FROM
  curated.trip
WHERE
  pickup_datetime >= '2024-01-15';

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

9 rows affected.

QUERY PLAN
Seq Scan on trip (cost=0.00..112674.80 rows=1679196 width=63) (actual time=278.945..517.581 rows=1678072 loops=1)
Filter: (pickup_datetime >= '2024-01-15 00:00:00'::timestamp without time zone)
Rows Removed by Filter: 1286552
Planning Time: 0.326 ms
JIT:
Functions: 2
"Options: Inlining false, Optimization false, Expressions true, Deforming true"
"Timing: Generation 0.316 ms, Inlining 0.000 ms, Optimization 0.396 ms, Emission 3.249 ms, Total 3.961 ms"
Execution Time: 590.036 ms


Lo importante es: cost=0.00..112674.80 rows=1679196 width=63 -- porcentaje de filas devueltas 56% 

# FASE 4 - Preparación para OLAP

Pensaremos en la parte de OLAP (modelo mental)

CURATED.zone no corrige solo modela (reduce columnas, define una dimension, no altera valores)
CURATED.vendor traduce códigos a significado, lógica SQL explicita

No hay joins prepara el camino:   curated.trip
    ├── vendor_id ──▶ curated.vendor
    ├── pulocation_id ──▶ curated.zone
    └── dolocation_id ──▶ curated.zone

curated.zone es un dataset externo, la unica función es proveer atributos descriptivos para poder hacer JOINs y denormalizar datos

curated.vendor es un lookup sintentico creado unicamente para mostrar como se agrega el contexto, traduce codigo, se enriquece una tabla de hechos

mapa mental: 
curated.trip   (OLTP-like, normalizada)
      │
      │ + zone (lookup real)
      │ + vendor (lookup sintético)
      ▼
VIEW / MV
modelo de lectura analítica

Comenzando a codificar: 

Primero vemos que en el csv yellow_tripdata_2024-01.csv hay dos campo PULocationID y DOLocationID esto será clave para la parte de taxi_zone_lookup.csv

desde el temrinal: wget https://d37ci6vzurychx.cloudfront.net/misc/taxi_zone_lookup.csv

Creación de la tabla postgres: taxi_zone_lookup (4 campos sin PKs sin FK y sin reglas)

In [1]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE TABLE IF NOT EXISTS raw.taxi_zone_lookup (
    location_id   INTEGER,
    borough       TEXT,
    zone_name     TEXT,
    service_zone  TEXT
);

bash: line 7: warning: here-document at line 1 delimited by end-of-file (wanted `SQL')


CREATE TABLE


In [2]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
\copy raw.taxi_zone_lookup FROM 'taxi_zone_lookup.csv'CSV HEADER;
SQL

COPY 265


Comprobación tabla

In [5]:
%%sql
SELECT
  *
FROM
  raw.taxi_zone_lookup
LIMIT
  10

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

10 rows affected.

location_id,borough,zone_name,service_zone
1,EWR,Newark Airport,EWR
2,Queens,Jamaica Bay,Boro Zone
3,Bronx,Allerton/Pelham Gardens,Boro Zone
4,Manhattan,Alphabet City,Yellow Zone
5,Staten Island,Arden Heights,Boro Zone
6,Staten Island,Arrochar/Fort Wadsworth,Boro Zone
7,Queens,Astoria,Boro Zone
8,Queens,Astoria Park,Boro Zone
9,Queens,Auburndale,Boro Zone
10,Queens,Baisley Park,Boro Zone


Vamos a crear una vista llamada curated.zone

In [6]:
%%sql
CREATE
OR REPLACE VIEW curated.zone AS
SELECT
  location_id,
  zone_name,
  borough
FROM
  raw.taxi_zone_lookup;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

++
||
++
++

Comprobamos la vista: 

In [7]:
%%sql
SELECT
  *
FROM
  curated.zone
LIMIT
  10

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

10 rows affected.

location_id,zone_name,borough
1,Newark Airport,EWR
2,Jamaica Bay,Queens
3,Allerton/Pelham Gardens,Bronx
4,Alphabet City,Manhattan
5,Arden Heights,Staten Island
6,Arrochar/Fort Wadsworth,Staten Island
7,Astoria,Queens
8,Astoria Park,Queens
9,Auburndale,Queens
10,Baisley Park,Queens


Creamos una vista llamada curated.vendor en este caso queremos una fila por vendor (duplicados) pasas de eventos a entidad. Por otra parte, traducimos a un significado humano el vendor_id ID_Técnico a vendor_name (nombre del vendor).

In [8]:
%%sql
CREATE
OR REPLACE VIEW curated.vendor AS
SELECT DISTINCT
  vendor_id,
  CASE vendor_id
    WHEN 1 THEN 'Creative Mobile'
    WHEN 2 THEN 'VeriFone'
    ELSE 'Unknown'
  END AS vendor_name
FROM
  curated.trip;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

++
||
++
++

Comprobamos vista: curated.vendor

In [9]:
%%sql
SELECT
  *
FROM
  curated.vendor

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

3 rows affected.

vendor_id,vendor_name
2,VeriFone
6,Unknown
1,Creative Mobile


Existen vendors que no conozco aún per no los oculto, revela problemas de calidad. En la documentación publica del dataset NYC Taxi & Limousine Commission (TLC) sale el significado de los IDs Vendors

En una capa dbt/gobernanza ya no existe el CASE. Será una tabla fuente oficial o un seed versionado (dim_vendor)

Las vista servirá por tanto para reducir joins, modelo mental, mejorar la legilibilidad de queries y para preparar el terreno para OLAP sin gobernanza aún.

Conclusión de SQL-First: “Con SQL puedes ir desde el CSV hasta un pseudo-OLAP funcional,
sin frameworks, sin herramientas externas, solo entendiendo el motor.”

MV Materialized View: 
Almacena resultados, se actualiza en intervalos, bajo demanda o automaticamente. Ideal para analitica. Es rapida en lectura, reduce carga de tablas OLTP pero ojo! puede estar los datos desfasados, requiere gestión de refresh y ocupa almacenamiento.

Curated.trip que es la tabla vamos a crear una materilizada view "vw_trip_denorm" se hace con psql no con pysql porque necesitamos esta herramienta para ser más rapido a la hora de procesar.

In [10]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE OR REPLACE VIEW curated.vw_trip_denorm AS
SELECT
    t.trip_id,
    DATE(t.pickup_datetime) AS pickup_date,
    EXTRACT(HOUR FROM t.pickup_datetime) AS pickup_hour,
    t.trip_distance,
    t.total_amount,
    v.vendor_name,
    z.zone_name  AS pickup_zone,
    z.borough   AS pickup_borough
FROM curated.trip t
LEFT JOIN curated.vendor v
       ON t.vendor_id = v.vendor_id
LEFT JOIN curated.zone z
       ON t.pulocation_id = z.location_id;

bash: line 16: warning: here-document at line 1 delimited by end-of-file (wanted `SQL')


CREATE VIEW


In [11]:
%%sql
SELECT
  *
FROM
  curated.vw_trip_denorm
LIMIT
  10

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

10 rows affected.

trip_id,pickup_date,pickup_hour,trip_distance,total_amount,vendor_name,pickup_zone,pickup_borough
3176581,2024-01-01,0.0,1.72,22.7,VeriFone,Penn Station/Madison Sq West,Manhattan
3176582,2024-01-01,0.0,1.8,18.75,Creative Mobile,Lenox Hill East,Manhattan
3176583,2024-01-01,0.0,4.7,31.3,Creative Mobile,Upper East Side North,Manhattan
3176584,2024-01-01,0.0,1.4,17.0,Creative Mobile,East Village,Manhattan
3176585,2024-01-01,0.0,0.8,16.1,Creative Mobile,SoHo,Manhattan
3176586,2024-01-01,0.0,4.7,41.5,Creative Mobile,Lower East Side,Manhattan
3176587,2024-01-01,0.0,10.82,64.95,VeriFone,LaGuardia Airport,Queens
3176588,2024-01-01,0.0,3.0,30.4,Creative Mobile,West Chelsea/Hudson Yards,Manhattan
3176589,2024-01-01,0.0,5.44,36.0,VeriFone,Midtown Center,Manhattan
3176590,2024-01-01,0.0,0.04,8.0,VeriFone,Greenwich Village North,Manhattan


No hemos hecho limpieza de datos y no se hace sobre RAW, solo las hemos cargado consolidando la información en una vista. Solo para consumo

el campo pickup_date damos formato fecha pero por preparación de que el analisis sea claro no por transformación de datos al igual que pickup_hour.

Los hechos no se negocian (no podemos perder filas, eventos)
El contexto se puede reconstruir después.

Es decir: Primero preservo eventos. El modelo se refina después.


Creamos una vista materializada "curated.mv_daily_zone" Prefiero perder detalle antes de perder estabilidad (Pensamiento OLAP)

In [15]:
%%sql
CREATE MATERIALIZED VIEW curated.mv_daily_zone AS
SELECT
    DATE(pickup_datetime) AS pickup_date,
    pulocation_id,
    COUNT(*)              AS trips,
    SUM(total_amount)     AS revenue
FROM curated.trip
GROUP BY
    DATE(pickup_datetime),
    pulocation_id;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

6966 rows affected.

++
||
++
++

In [16]:
%%sql
SELECT * FROM curated.mv_daily_zone

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

6966 rows affected.

pickup_date,pulocation_id,trips,revenue
2024-01-21,41,218,4247.97
2024-01-10,210,6,279.0
2024-01-31,14,10,524.95
2024-01-29,35,17,626.48
2024-01-20,79,4473,94645.55
2024-01-09,28,19,1050.00
2024-01-08,137,914,19275.47
2024-01-24,182,9,331.94
2024-01-15,50,499,10518.59
2024-01-04,117,12,560.35


RAW        → guardar
CURATED    → estructurar
VIEW       → explicar
MV         → acelerar

Diferencia: 
VIEW             → se recalcula siempre
MATERIALIZED VIEW→ se precalcula y persiste

Resumen:

RAW        → guardar
CURATED    → estructurar
VIEW       → explicar
MV         → acelerar

Si el analista escribe joins,
el modelo no existe.